Assign Roles with Active Record enum method

Andycutter opened the pull request:
### PT link
https://www.pivotaltracker.com/story/show/159732583

### User story
```
As a system owner,
In order to differentiate between the Users,
I would like to be able to apply different roles to Users.
```

### New Gems
`gem 'pundit'`
`  gem 'pundit-matchers', '~> 1.6.0'`

### What did we learn?
That active record 'enum' can be used to define a role in a single column in the DB by using an integer for the different roles
To use Pundit for authorization
To write unit tests for authorization

Andycutter committed:
add role with enum active record method, add 4 roles: user, subscriber, author and editor

Andycutter committed:
Merge branch 'Development' of https://github.com/CraftAcademy/D-pressed into roles_setup_enum

Andycutter committed:
add user roles with enum

Andycutter committed:
add test for role column

tochman commented on app/models/user.rb:
```diff
@@ -1,6 +1,14 @@
 class User < ApplicationRecord
+  after_initialize :set_default_role, :if => :new_record?
+  
+  enum role: [:user, :subscriber, :author, :editor]
```
IF you want to be in control of the numbers, you could do something like this:
```
enum role: [{user: 0}, {subscriber: 1}, {author: 10}, {editor: 5}]
```

Andycutter committed:
add feature file to test for users access to create articles page

Andycutter committed:
add custom numbers to roles

Andycutter committed:
rename user to basic_user to better distinguish

Andycutter committed:
redo migration

Andycutter committed:
fix syntax error for roles, rspec goes green now

Andycutter committed:
update factory to make it valid

Andycutter committed:
modify link name to not cause ambigious match

Andycutter committed:
resolve merge conflic from upstream Development

Andycutter committed:
add gem Pundit

Andycutter committed:
add pundit to application controller

Andycutter committed:
run migration and install pundit

Andycutter committed:
add unit test of user roles

Andycutter committed:
add test for all roles

Andycutter committed:
add gem pundit matchers to unit test policies

Andycutter committed:
add error notice for authorized error

Andycutter committed:
add authorize to @article

Andycutter committed:
clean ApplicationPolicy and move methods to article policy

Andycutter committed:
add polcies for index, show and create for now

Andycutter committed:
made feature, add specs to test and modify path_steps

Andycutter committed:
remove authorize from show, rename user to current_user in policies, remove == check in article_policy adn rspec went green

Andycutter committed:
typo

Andycutter committed:
add policy check for new instead of create, only authors and editors see the link to create article page
basic_user and subscriber get access denied if the try to access page

Andycutter committed:
add role to article creator in create article feature to make it pass with the new policies

Andycutter committed:
everything green but ALOT of false positives!

Andycutter committed:
add test for :new

Andycutter committed:
refactor user factory, modify features, add new step for page expectation

Andycutter committed:
remove binding.pry from features

Andycutter committed:
typo in email for subscriber

tochman commented on app/policies/article_policy.rb:
```diff
@@ -0,0 +1,18 @@
+  end
+
+  def show?
+    current_user.author? || current_user.editor? || current_user.subscriber?
```
````
!current_user.basic_user?
````

Andycutter committed:
refactor article policy to use less code,
remove redudant notices from views,
remove email from features and specs, since FactoryBot takes care of it now

Andycutter committed:
refactor page_path step

ESteinman reviewed (approved):
LGTM

Andycutter committed:
formatting and remove redudant empty lines

Andycutter committed:
merge conflicts from upstream Development resolved

Andycutter committed:
formatting

Andycutter committed:
modify feature to pass role authorization
